beefy-primitives/src/commitment.rs

```diff
@@ -95,6 +99,136 @@ impl<TBlockNumber, TPayload, TSignature> SignedCommitment<TBlockNumber, TPayload
+struct CompactSignedCommitment<TCommitment, TSignature> {
+	/// The commitment signatures, as is.
+	commitment: TCommitment,
+	/// A `Vec<u8>`, where each bit in a `u8` represents an `Option<TSignature>`
+	/// in the uncompressed GRANDPA validators' signatures.
+	signatures_from: BitField,
+	/// Length of the `Vec` containing uncompressed GRANDPA validators' signatures.
+	signatures_len: u32,
+	/// A `Vec` containing all `TSignature` values present in the uncompressed
+	/// GRANDPA validators' signatures.
+	signatures_compact: Vec<TSignature>,
+}
+
+impl<'a, TBlockNumber, TPayload, TSignature>
+	CompactSignedCommitment<&'a Commitment<TBlockNumber, TPayload>, &'a TSignature>
+where
+	TSignature: Encode,
+	TBlockNumber: Encode,
+	TPayload: Encode,
+{
+	/// Packs a `SignedCommitment` into the compressed `CompactSignedCommitment` format for efficient network transport.
+	fn pack(signed_commitment: &'a SignedCommitment<TBlockNumber, TPayload, TSignature>) -> Self {
+		let SignedCommitment { commitment, signatures } = signed_commitment;
+		let signatures_len = signatures.len() as u32;
+		let mut signatures_from: BitField = vec![];
+		let mut signatures_compact: Vec<&TSignature> = vec![];
+
+		for signature in signatures {
+			match signature {
+				Some(value) => signatures_compact.push(value),
+				None => (),
+			}
+		}
+
+		let bits: Vec<u8> = signatures.iter().map(|x| if x.is_some() { 1 } else { 0 }).collect();
+		let chunks = bits.chunks(CONTAINER_BIT_SIZE);
+		for chunk in chunks {
+			let mut iter = chunk.iter().copied();
+			let mut v = iter.next().unwrap() as u8;
+
+			for bit in iter {
+				v <<= 1;
+				v |= bit as u8;
+			}
+
+			signatures_from.push(v);
+		}
+
+		Self {
+			commitment,
+			signatures_from,
+			signatures_len,
+			signatures_compact,
+		}
+	}
+}
+
+impl<TBlockNumber, TPayload, TSignature> Encode for SignedCommitment<TBlockNumber, TPayload, TSignature>
+where
+	TSignature: Encode,
+	TBlockNumber: Encode,
+	TPayload: Encode,
+{
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+		let temp = CompactSignedCommitment::pack(self);
+		temp.using_encoded(f)
+	}
+}
+
+impl<TBlockNumber, TPayload, TSignature> CompactSignedCommitment<Commitment<TBlockNumber, TPayload>, TSignature>
+where
+	TBlockNumber: Decode,
+	TPayload: Decode,
+	TSignature: Decode,
+{
+	/// Unpacks a `CompactSignedCommitment` into the uncompressed `SignedCommitment` form.
+	fn unpack(
+		temporary_signatures: CompactSignedCommitment<Commitment<TBlockNumber, TPayload>, TSignature>,
+	) -> SignedCommitment<TBlockNumber, TPayload, TSignature> {
+		let CompactSignedCommitment {
+			commitment,
+			signatures_from,
+			mut signatures_len,
+			signatures_compact,
+		} = temporary_signatures;
+		let mut bits: Vec<u8> = vec![];
+		let last_block = signatures_len % CONTAINER_BIT_SIZE as u32;
+		for block in signatures_from {
+			let start_bit = if signatures_len > last_block {
+				0
+			} else {
+				CONTAINER_BIT_SIZE - last_block as usize
+			};
```

Comment:
I don't think this condition is correct, it does not depend on any of the loop items. 

Instead I'd suggest we take all elements from `signature_from` except for the last one which might not be full and only then we process the last item up to the number of bits defined in `last_block`.